File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.py
class Solution:
    def maxProfit(self, prices: List[int]) -> int:
        n = len(prices)
        dp = [0] * n

        for _ in range(2): 
            diff = float('-inf')
            prev = 0
            for i, price in enumerate(prices):
                diff = max(diff, prev - price)
                prev = dp[i]
                if i > 0:
                    dp[i] = max(dp[i-1], diff + price)
                else:
                    dp[i] = max(dp[i], diff + price)

        return dp[-1]

File: 0123-best-time-to-buy-and-sell-stock-iii/0123-best-time-to-buy-and-sell-stock-iii.cpp
class Solution {
public:
    int maxProfit(vector<int>& prices) {
        int min1 = INT_MIN, min2 = INT_MIN;
        int max1 = 0, max2 = 0;
        for(auto& price : prices){
            max2 = max(max2, price + min2);
            min2 = max(min2, max1 - price);
            max1 = max(max1, min1 + price);
            min1 = max(min1, -price);
        }
        return max2;
    }
};